### PR TITLE
Support enriched metadata

### PR DESCRIPTION
**Pull request by brettcannon**
Take raw metadata and wrap data as appropriate with more enriched representations (e.g., `requirements.Requirement` for `Requires-Dist`).

Part of #570

### CODE REVIEW

**Comment by brettcannon** [comment]
Before I add documentation and take this out of being a draft PR (assuming folks agree with the approach), there are some open questions I would like help answering (search for `TODO` in the PR):

1. Should this check `unparsed` for valid keys?
2. Check that fields are specified in a valid metadata version? Should that be done per-field, or should it be done in a more over-arching validate() method?
3. `description` can only be specified in the body in metadata 2.1; do we need to check for that explicitly?
4. Are the various parts of `description_content_type` case-insensitive?

**Comment by brettcannon** [comment]
And FYI the failures are code coverage and Python 3.7 & 3.8 not liking `Generic[T]`.

**Comment by brettcannon** [comment]
FYI if people aren't available to provide feedback before the end of the month (e.g., @pradyunsg ), then I'll close the PR and do I need over in mousebender directly to unblock myself.

File: src/packaging/metadata.py
Comment (Secrus) [inline src/packaging/metadata.py:576]: While `from_email` is obvious for those who know about how `PKG-INFO`/`METADATA` files are constructed (email headers), as an end-user I would expect something like `from_metadata_file(file)` or something similar. However, I can understand if this could be too much. Basically, I would expect `packaging.metadata` to be a semi-drop-in replacement for [`pkginfo`](https://pythonhosted.org/pkginfo/)

Comment (brettcannon) [reply src/packaging/metadata.py:576]: I'm not totally opposed to a nicer name, but it matches https://packaging.pypa.io/en/stable/metadata.html#packaging.metadata.parse_email currently.

As for the the `file` argument, we are specifically taking a [sans-I/O](https://sans-io.readthedocs.io/) approach to this API (and this project overall), so we won't have an API that takes a file path (closest would be for a file-like object, but even then I don't know if it's worth it since people could be getting this info via an HTTP request to PyPI).

Comment (Secrus) [reply src/packaging/metadata.py:576]: I understand the sans-I/O part, it makes total sense. I would still go for a nicer name though. However, I have no idea for a good one. `from_metadata_content(content: str)`? Doesn't sound right. 

On the other hand, most users would probably be knowledgeable enough to understand the `email` part of the name. Especially if it matches the current API.

File: src/packaging/metadata.py
Comment (pradyunsg) [inline src/packaging/metadata.py:434]: _sigh_

File: src/packaging/metadata.py
Comment (pradyunsg) [inline src/packaging/metadata.py:507]: nit: Consider pulling this out into a global, around the `_MetadataVersion` global.

**Review by pradyunsg — APPROVED** [review]
LGTM, with mostly nit-picks. Appreciate your patience around this, while I basically vanished for a month from OSS unannounced. :)

File: docs/utils.rst
Comment (brettcannon) [reply docs/utils.rst:17]: Definitely! I don't know why I didn't (I probably thought I couldn't due to backwards-compatibility).

File: tests/test_metadata.py
Comment (brettcannon) [reply tests/test_metadata.py:290]: It's definitely an open question. I'm personally fine with being strict and saying folks are expected to use raw metadata parsing in those instances.

Comment (dstufft) [reply tests/test_metadata.py:290]: We can always start with it being invalid and loosen it later if it turns out there is a lot of breakage.

I have vague memories that some versions of wheel was emitting it though, so it may be a lot of people. On the other hand, I think my original PR (I haven't checked this one) also failed on unknown keys, and a bunch of people are currently emitting the License PEP metadata that hasn't landed... I think hatch does that? 

My hope is for this to be as strict as we can though, so I can stick it in the upload flow to warehouse and stop the bleeding :)

**Comment by pradyunsg** [comment]
> Should this check unparsed for valid keys?

I think so -- you mean that `unparsed` is empty, or like, `X-foo` style keys are permitted or so?

> Check that fields are specified in a valid metadata version? Should that be done per-field, or should it be done in a more over-arching validate() method?

It would be nice to do that -- TBH, I wouldn't mind having all the validation happen within the `from_raw` step. A `validate()` works too, as does doing it per-field.

> description can only be specified in the body in metadata 2.1; do we need to check for that explicitly?

I don't think we _need_ to, but it would be good to. Not a blocking concern from my PoV.

> Are the various parts of description_content_type case-insensitive?

The `text/plain` parts? No. The parameters? Yes. I'm saying this based on https://stackoverflow.com/a/12869287/1931274.

File: src/packaging/metadata.py
Comment (dstufft) [inline src/packaging/metadata.py:580]: I would definitely check for unparsed keys and consider it a validation fail. If someone wants to ignore those keys it's a two line function, they can do a little copying.

File: src/packaging/metadata.py
Comment (dstufft) [inline src/packaging/metadata.py:587]: If I'm reading this correctly, I don't think we want to normalize the name here. The core metadata spec doesn't require normalization in general, just normalization prior to comparison, and I know from PyPI that people get very attached to the formatting of their name.

Should still validate that it's a valid value though.

**Comment by brettcannon** [comment]
> > Should this check unparsed for valid keys?
> 
> I think so -- you mean that `unparsed` is empty, or like, `X-foo` style keys are permitted or so?

I was thinking more the latter, but I think @dstufft wants the former.

File: src/packaging/metadata.py
Comment (brettcannon) [reply src/packaging/metadata.py:587]: I take it you're thinking about having Warehouse get the name via this code and thus changing its format here would be bad? Otherwise my thinking was this avoided someone messing up the comparison later on by forgetting to normalize (i.e., normalize on the edges of your code before the data gets into your business logic).

**Comment by dstufft** [comment]
Yea, I don't think unknown data should be dropped silently. I think it's a common source of problems. When I was trying my original branch, which did that, against the `METADATA` files that I pulled down from PyPI, that caught a fair number of weird and broken metadata.

File: src/packaging/metadata.py
Comment (dstufft) [reply src/packaging/metadata.py:587]: That's my plan yea, though I don't think it's *just* Warehouse because it can affect things like `pip show` and stuff too, but I suspect the people who care will primarily care about what Warehouse does, so it's *mostly* warehouse I guess.

At one point someone.. might have been you, talked about adding a `metadata.display_name()`, and that would honestly be fine with me too, though I have a slight preference for it just not canonicalizing because it just feels a little off to me? But that may just be my weirdness so if others feel differently I'm happy as long as I can still get the valid, but not canonicalized name.

Could always do something gross:

```python
class Name(str):
    def __eq__(self, other):
        return canoncalize_name(self) == canonicalize_name(other)
        
    # etc
```

Anyways, I'm cool with whatever you think here. I certainly wouldn't block it over that. That's just what I expect people will complain about.

**Comment by brettcannon** [comment]
My TODO list:

- [ ] Check for content-type in a case-insensitive way
- [ ] Don't normalize `name`, just check that it's valid
- [ ] Drop `2.0` as a valid metadata version
- [ ] Address remaining review comments
- [ ] Allow for full validation via `from_raw()` (probably calls for `ExceptionGroup`)
- [ ] Anything unparsed is a validation failure
- [ ] Support Python 3.8
- [ ] Get full code coverage
- [ ] Support Python 3.7? (will probably have reached EOL at this point, so check if we are still supporting it)

**Comment by brettcannon** [comment]
The code is now at 100% coverage and works in >= 3.8.

@pradyunsg or @dstufft , did you want to review the code now, or wait until I have written the docs?

**Comment by brettcannon** [comment]
Docs are in! https://github.com/pypa/packaging.python.org/pull/1283 will let me go back in and add links to the core metadata fields that I couldn't link to.

All of CI is now passing, so does anyone want to give this a once-over?

**Review by pradyunsg — APPROVED** [review]
Did a quick skim, LGTM!

**Comment by brettcannon** [comment]
@pradyunsg thanks for the skim!

I'll give @dstufft some time to look it over before I merge.

**Comment by brettcannon** [comment]
What do we want to do about core metadata 1.1 and its fields that **only** existed in that version (i.e., `requires`, `provides`, and `obsoletes`)? Right now the fields are considered a validation failure since there just isn't a concept of those fields in the class. Should I just assign that data to the roughly corresponding core metadata 1.2 fields? Should I just not support core metadata 1.0 and 1.1 (or just 1.1)?

**Comment by dstufft** [comment]
I don't think we've ever *removed* a field? The PEP for Metadata 1.2 or 2.1 never said that those fields were removed, just deprecated.

**Comment by dstufft** [comment]
It looks like https://github.com/pypa/packaging.python.org/pull/386 just missed documenting those fields, because PEP 345 doesn't document them (other than as a footnote to say they're deprecated), and AFAIK most or all existing implementations of metadata reading that attempt to be comprehensive (so not things trying to get a single field or something) support them still.

**Review by dstufft — APPROVED** [review]
This approach looks good to me, I think it covers all the bases/invariants that I was trying to handle in my original PR, looks great!

**Comment by brettcannon** [comment]
> I don't think we've ever _removed_ a field? The PEP for Metadata 1.2 or 2.1 never said that those fields were removed, just deprecated.

Ah, OK. I have been following the spec as "the spec", and so figured the fields were gone. I will add the missing ones in as `str`-only fields and that should take care of things! I'll also open an issue on packaging.python.org about the missing fields.

**Comment by brettcannon** [comment]
Actually, Donald reported the doc bug in https://github.com/pypa/packaging.python.org/issues/1107 and there's a PR at https://github.com/pypa/packaging.python.org/pull/1138 waiting on Donald's review. 😅

**Comment by dstufft** [comment]
> Donald reported the doc bug in https://github.com/pypa/packaging.python.org/issues/1107

Welp.

**Comment by brettcannon** [comment]
I wrote a script that tried to pull a `METADATA` file from every project on PyPI. The issues that I saw, in order of most to least frequent, were:

1. Using core metadata version 2.0
2. Improper description content type
3. Invalid extras name (people definitely want a name to specify development-only requirements)
4. Multi-line summaries
5. Repeated fields (in one case because they messed up their `Description` formatting and made it a blank line and then followed with the whole description in the body)

I need to make a change to make it more descriptive what triggered when the name is invalid (since it's reusing `utils.canonicalize_name()` it has a generic exception message), but otherwise I think that's the only change left!

**Comment by brettcannon** [comment]
OK, I updated the appropriate code so that all failed field validations raise `InvalidMetadata` (with appropriate `__cause__` values). That way you only have to worry about catching a single exception type instead of having to know exactly what code was used to perform the validation or transformation of code.

And with that, I think this PR is ready! I'm waiting for CI to finish, so hopefully I can merge this tomorrow.

File: CHANGELOG.rst
Comment (atugushev) [inline CHANGELOG.rst:10]: Spotted some minors https://github.com/pypa/packaging/pull/709

**Comment by henryiii** [comment]
Python 3.8 is now supported, not sure how `cached_property` would be useful here (other than for `raw_name`, perhaps?

**Comment by brettcannon** [comment]
> Python 3.8 is now supported, not sure how `cached_property` would be useful here (other than for `raw_name`, perhaps?

Maybe. At least for `Metadata`, that data storage is private and not there's no way to mutate it from the properties, so caching the results _should_ work. I think it's just a question of whether it lowers code maintenance or not.

**Pull request merged**
